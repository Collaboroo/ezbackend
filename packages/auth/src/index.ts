import 'fastify-passport';
export * from './auth';
export * from './ez-user';
export * from './providers';
