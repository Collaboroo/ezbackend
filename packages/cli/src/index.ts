import { Command } from "commander";
import { sync as readPackageUpSync } from "read-pkg-up";
import initiate from "./initiate";

//URGENT TODO: Make dependencies core dependencies not dev dependencies during installation
//URGENT TODO: Add @ezbackend/core to dependencies

const pkg = readPackageUpSync({ cwd: __dirname })?.packageJson;

const program = new Command();

program
  .command("init")
  .description("Initialize ezbackend into your project.")
  .option("-f --force", "Force add ezbackend")
  .action((options) => initiate(options, pkg));

program.parse(process.argv);

