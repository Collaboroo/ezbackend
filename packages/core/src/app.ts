import {
    kApp,
    kInstance,
    kScope,
} from './symbols'
import avvio, { Avvio, mixedInstance, Plugin } from 'avvio'

export type PluginType = Plugin<unknown, AppInstance>
export type Lifecycle =
    '_preInit' |
    '_init' |
    '_postInit' |
    '_preHandler' |
    '_handler' |
    '_postHandler' |
    '_preRun' |
    '_run' |
    '_postRun'

export enum PluginScope {
    PARENT,
    DEFAULT
}

export const LIFECYCLE: Array<Lifecycle> = [
    '_preInit',
    '_init',
    '_postInit',
    '_preHandler',
    '_handler',
    '_postHandler',
    '_preRun',
    '_run',
    '_postRun'
]

export class AppInstance {
    [kApp]: App
}

export type Override = Avvio<AppInstance>['override']

export type Overrides = {[name:string]: Avvio<unknown>['override']}

//URGENT TODO: Add types
//TODO: Added safety for overriding instance variables?
export class App {
    protected _parent: App | undefined
    protected _apps: Map<string, App>
    protected _preInit: Map<string, Plugin<any, any>>
    protected _init: Map<string, Plugin<any, any>>
    protected _postInit: Map<string, Plugin<any, any>>
    protected _preHandler: Map<string, Plugin<any, any>>
    protected _handler: Map<string, Plugin<any, any>>
    protected _postHandler: Map<string, Plugin<any, any>>
    protected _preRun: Map<string, Plugin<any, any>>
    protected _run: Map<string, Plugin<any, any>>
    protected _postRun: Map<string, Plugin<any, any>>
    protected _instance: Avvio<AppInstance>
    protected _name: string
    protected _scope: PluginScope
    protected _overrides: Overrides

    //TODO: Make this follow the convention I guess
    opts: any

    //TODO: Add tests for localdata
    //TOOD: Add test for obtaining parent
    //TODO: Replace errors with better errors

    constructor() {
        this._apps = new Map()
        this._preInit = new Map()
        this._init = new Map()
        this._postInit = new Map()
        this._preHandler = new Map()
        this._handler = new Map()
        this._postHandler = new Map()
        this._preRun = new Map()
        this._run = new Map()
        this._postRun = new Map()
        this._instance = avvio(new AppInstance(), { autostart: false })
        this._name = 'Root'
        this._scope = PluginScope.DEFAULT
        this._overrides = {}
        this.opts = {}
    }

    get apps() { return this._apps }
    get init() { return this._init }
    get handler() { return this._handler }
    get run() { return this._run }
    get instance() { return this._instance }
    get name() { return this._name }
    get scope() { return this._scope }
    get parent() { return this._parent }
    get overrides() {return this._overrides}

    set name(newName: string) {
        //Should we prevent setting the name more than once?
        this._name = newName
    }

    set scope(newScope: PluginScope) {
        this._scope = newScope
    }

    setPreInit(funcName: string, plugin: Plugin<any, any>) { this.setHook('_preInit', funcName, plugin) }
    setInit(funcName: string, plugin: Plugin<any, any>) { this.setHook('_init', funcName, plugin) }
    setPostInit(funcName: string, plugin: Plugin<any, any>) { this.setHook('_postInit', funcName, plugin) }
    setPreHandler(funcName: string, plugin: Plugin<any, any>) { this.setHook('_preHandler', funcName, plugin) }
    setHandler(funcName: string, plugin: Plugin<any, any>) { this.setHook('_handler', funcName, plugin) }
    setPostHandler(funcName: string, plugin: Plugin<any, any>) { this.setHook('_postHandler', funcName, plugin) }
    setPreRun(funcName: string, plugin: Plugin<any, any>) { this.setHook('_preRun', funcName, plugin) }
    setRun(funcName: string, plugin: Plugin<any, any>) { this.setHook('_run', funcName, plugin) }
    setPostRun(funcName: string, plugin: Plugin<any, any>) { this.setHook('_postRun', funcName, plugin) }


    removeHook(lifecycle: Lifecycle, funcName: string) {
        if (!this[lifecycle].has(funcName)) {
            throw `${funcName} does not exist in ${lifecycle}`
        }
        //Override the plugin name
        this[lifecycle].delete(funcName)

    }

    setHook(lifecycle: Lifecycle, funcName: string, plugin: Plugin<any, any>) {
        if (this[lifecycle].has(funcName)) {
            throw `${funcName} already declared for ${lifecycle}`
        }
        //Override the plugin name
        Object.defineProperty(plugin, 'name', { value: funcName })
        this[lifecycle].set(funcName, plugin)

    }

    _setParent(app: App) {
        if (this._parent === undefined) {
            this._parent = app
        } else {
            throw `App's parent has already been set, parents can only be set once`
        }
    }



    addApp(name: string, newApp: App, opts: any = {}) {
        if (name in this._apps || Object.values(this._apps).indexOf(newApp) !== -1) {
            throw (`Child app ${name} already exists`)
        }
        newApp.name = name
        newApp.opts = opts
        newApp._setParent(this)
        this._apps.set(name, newApp)
        
    }

    getApp(name) {
        return this.apps.get(name)
    }

    getHookPlugin(lifecycle: Lifecycle): undefined | PluginType {

        //URGENT TODO: Plugin edge test case where child has hook but parent does not
        const childPlugins = Array.from(this.apps.values()).reduce((previousPlugins, app) => {
            const childPlugin = app.getHookPlugin(lifecycle)
            if (childPlugin) {
                previousPlugins.push(childPlugin)
            }
            return previousPlugins
        }, [])

        if (this[lifecycle].size === 0 && childPlugins.length === 0) {
            return undefined
        }

        const appFunc: PluginType = async (instance, opts) => {
            this[lifecycle].forEach(async pluginFunc => {
                /*
                Take note he who dares venture here:
                The base code compiles into js, so everything is Function there
                However when the user writes his own code, the functions can be AsyncFunction
                So the prototype may be empty
                Which is why we need to define it if necessary
                TODO: Add test case for above edge case
                */
                if (pluginFunc.prototype === undefined) {
                    pluginFunc.prototype = {}
                }
                //Make sure plugins run on the parent scope
                pluginFunc.prototype[kScope] = PluginScope.PARENT
                //TODO: Test case/throw error to ensure that kApp is set for ALL plugins
                pluginFunc.prototype[kApp] = this
                instance.use(pluginFunc, opts)
            })
            childPlugins.forEach(childPlugin => {
                instance.use(childPlugin, opts)
            })
        }

        appFunc.prototype[kApp] = this
        appFunc.prototype[kScope] = this.scope
        //Override the encapsulating function's name
        Object.defineProperty(appFunc, 'name', { value: this.name })

        return appFunc
    }

    async start(opts?: any) {
        if (this._instance.started) {
            throw `App has been started already!`
        }
        //Configure encapsulation
        this._instance.override = this.override
        //Recursively load plugins
        LIFECYCLE.forEach(lifecycle => {
            const lifecyclePlugin = this.getHookPlugin(lifecycle)
            if (lifecyclePlugin) {
                Object.defineProperty(lifecyclePlugin, 'name', { value: lifecycle })
                this._instance.use(lifecyclePlugin, opts)
            }
        })
        //Start the instance
        this._instance.start()
        await this._instance.ready()
    }

    setCustomOverride(varName:string, override: Avvio<unknown>['override']) {
        this._overrides[varName] = override
    }

    protected override: Override = (old, fn, opts) => {
        let newInstance: mixedInstance<AppInstance>
        const parentApp = fn.prototype[kApp] ?? this
        //Check plugin scope
        if (fn.prototype[kScope] === PluginScope.DEFAULT) {
            //Uses new scope
            if (parentApp[kInstance] === undefined) {
                parentApp[kInstance] = Object.create(old)
                Object.entries(this.overrides).forEach(([varName,override]) => {
                    if (parentApp[kInstance][varName] === undefined) {
                        //Don't do anything if variable has not been defined yet
                        return
                    }
                    parentApp[kInstance][varName] = override(
                        parentApp[kInstance][varName],
                        async (instance,opts) => {},
                        parentApp.opts
                    )
                })
            }
            newInstance = parentApp[kInstance]
        } else if (fn.prototype[kScope] === PluginScope.PARENT) {
            //Uses old scope
            newInstance = old
        } else {
            throw `Function scope is not defined for ${fn}`
        }
        newInstance[kApp] = parentApp
        return newInstance
    }
}