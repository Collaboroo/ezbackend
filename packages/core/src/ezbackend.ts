import path from "path";
import fs from "fs";
import avvio from "avvio";

//TODO: Make it not unknown
export type IEzbPlugin = avvio.Plugin<unknown, EzBackend>;

export interface IEzbConfig {
  plugins: Array<string>;
  server?: unknown;
  orm?: unknown;
  port?: number;
  entryPoint?: string;
  connectionURI?: string;
}

export type IEzbPlugins = {
  preInit: Array<IEzbPlugin>;
  init: IEzbPlugin | null;
  postInit: Array<IEzbPlugin>;
  preHandler: Array<IEzbPlugin>;
  handler: IEzbPlugin | null;
  postHandler: Array<IEzbPlugin>;
  preRun: Array<IEzbPlugin>;
  run: IEzbPlugin | null;
  postRun: Array<IEzbPlugin>;
};

export class EzBackend {
  plugins: IEzbPlugins;

  private static instance: EzBackend;
  private static manager: avvio.Avvio<EzBackend>;

  //TODO: Think of why its not throuwing error when the type passed is not exactly an ezbplugin
  constructor() {
    this.plugins = {
      preInit: [],
      init: (ezb, opts, cb) => {
        cb();
      },
      postInit: [],
      preHandler: [],
      handler: (ezb, opts, cb) => {
        cb();
      },
      postHandler: [],
      preRun: [],
      run: (ezb, opts, cb) => {
        cb();
      },
      postRun: [],
    };
  }

  public static initializeApp() {
    if (!EzBackend.instance) {
      EzBackend.instance = new EzBackend();
      EzBackend.manager = avvio(EzBackend.app());
    }
  }

  public static app(): EzBackend {
    EzBackend.initializeApp();
    return EzBackend.instance;
  }

  public static async start(configPath?: string) {
    let customConfigs: IEzbConfig | undefined;

    const customConfigPath =
      configPath ?? path.resolve(process.cwd(), ".ezb/config.ts");

    if (fs.existsSync(customConfigPath)) {
      customConfigs = require(customConfigPath).default;
      customConfigs.plugins.forEach((pluginName) => {
        require(pluginName);
      });
    }

    //LOAD PLUGINS FROM CONFIG
    //TODO: Allow changing of config path, or default config if none

    //TODO: Error handling for wrong format of config.ts

    //TODO: Figure out how to pass in options at top level
    EzBackend.manager.use((ezb, opts, cb) => {
      //URGENT TODO: Error handling when plugin doesnt work
      const plugins = ezb.plugins;
      

      plugins.preInit.forEach((plugin) => {
        ezb.use(plugin, customConfigs);
      });

      ezb.use(plugins.init, customConfigs);

      plugins.postInit.forEach((plugin) => {
        ezb.use(plugin, customConfigs);
      });

      plugins.preHandler.forEach((plugin) => {
        ezb.use(plugin, customConfigs);
      });

      ezb.use(plugins.handler, customConfigs);

      plugins.postHandler.forEach((plugin) => {
        ezb.use(plugin, customConfigs);
      });

      plugins.preRun.forEach((plugin) => {
        ezb.use(plugin, customConfigs);
      });

      ezb.use(plugins.run, customConfigs);

      plugins.postRun.forEach((plugin) => {
        ezb.use(plugin, customConfigs);
      });
      cb();
    });


    EzBackend.manager.start();
    await EzBackend.manager.ready();
    return;
  }
}
