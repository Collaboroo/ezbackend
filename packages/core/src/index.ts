type IEzbPlugin = () => void;

export type IEzbPlugins = {
  preInit: Array<IEzbPlugin>;
  init: IEzbPlugin | null;
  postInit: Array<IEzbPlugin>;
  preHandler: Array<IEzbPlugin>;
  handler: IEzbPlugin | null;
  postHandler: Array<IEzbPlugin>;
  preRun: Array<IEzbPlugin>;
  run: IEzbPlugin | null;
  postRun: Array<IEzbPlugin>;
};

export class EzBackend {
  plugins: IEzbPlugins;

  private static instance: EzBackend;

  constructor() {
    this.plugins = {
      preInit: [],
      init: () => {},
      postInit: [],
      preHandler: [],
      handler: () => {},
      postHandler: [],
      preRun: [],
      run: () => {},
      postRun: [],
    };
  }

  public static initializeApp() {
    if (!EzBackend.instance) {
      EzBackend.instance = new EzBackend();
    }
  }

  public static app(): EzBackend {
    EzBackend.initializeApp();
    return EzBackend.instance;
  }

  public static start(): void {
    const ezb = EzBackend.app();
    const plugins = ezb.plugins;

    plugins.preInit.forEach((plugin) => {
      plugin();
    });

    plugins.init();

    plugins.postInit.forEach((plugin) => {
      plugin();
    });

    plugins.preHandler.forEach((plugin) => {
      plugin();
    });

    plugins.handler();

    plugins.postHandler.forEach((plugin) => {
      plugin();
    });

    plugins.preRun.forEach((plugin) => {
      plugin();
    });

    plugins.run();

    plugins.postRun.forEach((plugin) => {
      plugin();
    });
  }
}
