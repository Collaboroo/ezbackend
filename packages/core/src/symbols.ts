export const kApp = Symbol('ezb.app')
export const kInstance = Symbol('ezb.instance')
export const kScope = Symbol('ezb.scope')