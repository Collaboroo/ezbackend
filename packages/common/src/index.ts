import { IOptions } from "./definitions";
import { IEzbConfig } from "@ezbackend/core";
import { fastify } from "fastify";
import path from "path";
import { EzBackend } from "./definitions";
import { mixedInstance } from "avvio";
import { createConnection } from "typeorm";
import { RouteOptions } from "fastify";
import { APIGenerator } from "./models";
import {convert} from './models/typeorm-json-schema'

const ezb = EzBackend.app();

//Configure defaults
ezb.plugins.init = async (
  ezb: mixedInstance<EzBackend>,
  opts: IEzbConfig & IOptions,
  cb
) => {
  ezb.orm = await createConnection(opts.orm);
  ezb.server = fastify(opts.server);
  cb();
};

ezb.plugins.handler = (ezb: mixedInstance<EzBackend>, opts: IEzbConfig, cb) => {
  const customEzbPath =
    opts.entryPoint ?? path.join(process.cwd(), ".ezb/index.ts");
  require(customEzbPath);
  ezb.models.forEach((model) => {
    const repo = ezb.orm.getRepository(model);

    //LEFT OFF
    // const metaData = ezb.orm.getMetadata(model)
    // console.log(convert(metaData))

    const generator = new APIGenerator(repo, { prefix: repo.metadata.name });
    generator.generateRoutes();
  });
  cb();
};

ezb.plugins.run = async (
  ezb: mixedInstance<EzBackend>,
  opts: IEzbConfig,
  cb
) => {
  await ezb.server.listen(opts.port, function (err, address) {
    if (err) {
      console.error(err);
      process.exit(1);
    }
  });
  cb();
};

export { EzBackend } from "./definitions";
export { EzModel, response, APIGenerator } from "./models";
