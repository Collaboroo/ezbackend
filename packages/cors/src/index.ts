import { EzBackend } from "@ezbackend/core";
import { mixedInstance } from "avvio";

const ezb = EzBackend.app()

interface IOptions {
    cors?: any
}

//TODO: Log warning if the cors are not properly set up
ezb.plugins.postInit.push((ezb: mixedInstance<EzBackend>,opts:IOptions,cb)=> {
    ezb.server.register(require('fastify-cors'), opts.cors)
    cb()
})