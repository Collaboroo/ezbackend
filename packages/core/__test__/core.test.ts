import { describe, it, expect } from "@jest/globals";
import { App, PluginScope } from "../src";

describe("Default Behaviour", () => {
  it("Plugins run in order", async () => {
    const app = new App()
    const arr: Array<number> = [];
    app.setPreInit('preInit', async (instance, opts) => {
      arr.push(1);
    })
    app.setInit('init', async (instance, opts) => {
      arr.push(2);
    })
    app.setPostInit('postInit', async (instance, opts) => {
      arr.push(3);
    })
    app.setPreHandler('preHandler', async (instance, opts) => {
      arr.push(4);
    })
    app.setHandler('handler', async (instance, opts) => {
      arr.push(5);
    })
    app.setPostHandler('postHandler', async (instance, opts) => {
      arr.push(6);
    })
    app.setPreRun('preRun', async (instance, opts) => {
      arr.push(7);
    })
    app.setRun('run', async (instance, opts) => {
      arr.push(8);
    })
    app.setPostRun('postRun', async (instance, opts) => {
      arr.push(9);
    })
    await app.start()

    expect(arr).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9]);
  });

  it("Plugins in sub app run in order", async () => {
    const app = new App()
    const subApp = new App()
    const arr: Array<number> = [];
    app.setInit('init', async (instance, opts) => {
      arr.push(1);
    })
    app.setHandler('handler', async (instance, opts) => {
      arr.push(2);
    })
    app.setRun('run', async (instance, opts) => {
      arr.push(3);
    })
    subApp.setInit('sub_init', async (instance, opts) => {
      arr.push(1.5)
    })
    app.addApp('subApp', subApp)
    await app.start()

    expect(arr).toEqual([1, 1.5, 2, 3]);
  })

  it("Encapsulation allows access to only parents", async () => {
    const app = new App()
    const subApp = new App()
    const subApp2 = new App()

    app.addApp('subApp', subApp)
    app.addApp('subApp2', subApp2)

    app.setInit('init', async (instance, opts) => {
      instance.rootVar = 'rootVar'
    })

    subApp.setInit('init', async (instance, opts) => {
      instance.subApp1Var = 'subApp1Var'
      expect(instance.rootVar).toEqual('rootVar')
      expect(instance.subApp2Var).toEqual(undefined)
    })

    subApp2.setInit('init', async (instance, opts) => {
      instance.subApp2Var = 'subApp2Var'
      expect(instance.rootVar).toEqual('rootVar')
      expect(instance.subApp1Var).toEqual(undefined)
    })

    await app.start()
  })

  it("An app should not be able to have 2 parents", async () => {
    const app1 = new App()
    const app2 = new App()
    const childApp = new App()

    let errored

    app1.addApp(childApp)
    try {
      app2.addApp(childApp)
    } catch (e) {
      expect(e).toMatchSnapshot()
      errored = true
    }

    expect(errored).toBe(true)
  })

  it("addApp syntax must be strictly followed", () => {

    let errored
    const app1 = new App()
    try {
      //@ts-ignore
      app1.addApp("randomString", "randomString", {})
    } catch (e) {
      expect(e).toMatchSnapshot()
      errored = true
    }

    expect(errored).toBe(true)


  })

  it("Encapsulation allows access scope of previous lifecycle hooks", async () => {
    const app = new App()
    const subApp = new App()
    const subApp2 = new App()

    app.setInit('init', async (instance, opts) => {
      instance.rootVar = 'rootVar'
    })

    subApp.setInit('init', async (instance, opts) => {
      instance.subApp1Var = 'subApp1Var'
    })

    subApp.setHandler('handler', async (instance, opts) => {
      expect(instance.rootVar).toEqual('rootVar')
      expect(instance.subApp1Var).toEqual('subApp1Var')
      expect(instance.subApp2Var).toEqual(undefined)
    })

    subApp2.setInit('init', async (instance, opts) => {
      instance.subApp2Var = 'subApp2Var'
    })
    app.addApp('subApp', subApp)
    app.addApp('subApp2', subApp2)
    await app.start()
  })

  it("Using the PARENT scope should allow parents to change encapsulated children variables", async () => {
    const app = new App()
    const subApp = new App()

    subApp.scope = PluginScope.PARENT

    app.addApp(subApp)

    subApp.setInit("Create Variable", async (instance, opts) => {
      instance.myChildVar = "Hello"
    })

    app.setHandler("Check access to child var", async (instance, opts) => {
      expect(instance.myChildVar).toBe("Hello")
    })

    await app.start()


  })

  it("Two child apps should not have the same name", async () => {
    const app = new App()
    const child1 = new App()
    const child2 = new App()

    let errored = false

    app.addApp("child", child1)
    try {
      app.addApp("child", child2)
    } catch (e) {
      expect(e).toMatchSnapshot()
      errored = true
    }

    expect(errored).toBe(true)

  })

  it("Should be able to get app by name", async () => {
    const app = new App()
    const child1 = new App()

    app.addApp("child", child1)

    const theChild = app.getApp("child")

    expect(theChild).toBe(child1)

  })

  it("Should be able to add plugins with undefined prototypes, which occurs in compiled JS code", async () => {
    const app = new App()

    const myFunc = async (instance: any, opts: any) => { }

    //@ts-ignore
    myFunc.prototype = undefined

    app.setInit("MyFunc", myFunc)

    await app.start()


  })

  describe("Hooks", () => {
    let app: App
    let flag = false
    beforeAll(async () => {
      app = new App()
      app.setInit("Set Flag", async (instance, opts) => {
        flag = true
      })
    })

    it("Setting a hook twice should throw an error", () => {
      let errored = false
      try {
        app.setInit("Set Flag", async (instance, opts) => {
          flag = true
        })
      } catch (e) {
        expect(e).toMatchSnapshot()
        errored = true
      }

      expect(errored).toBe(true)

    })
    it("Removing a hook by name should work", () => {
      app.removeHook("_init", "Set Flag")

      app.start()

      expect(flag).toBe(false)
    })

    it("Removing a non-existent hook should throw an error", () => {
      let errored = false
      try {
        app.removeHook("_handler", "non-existent")
      } catch (e) {
        expect(e).toMatchSnapshot()
        errored = true
      }

      expect(errored).toBe(true)


    })


  })


  test.todo("Running app.start() twice should return an EzError, not a normal error")
  test('Running app.start() twice should error', async () => {
    const app = new App()
    await app.start()
    let errored = false
    try {
      await app.start()
    } catch (e) {
      errored = true
    }
    expect(errored).toBe(true)
  })

  describe("addApp overloads should function as expected", () => {
    it("Just app", async () => {
      const app = new App()
      const subApp = new App()
      const arr: Array<number> = [];
      app.setInit('init', async (instance, opts) => {
        arr.push(1);
      })
      app.addApp(subApp)
      await app.start()

      expect(arr).toEqual([1]);
    })

    it("Just app and Opts", async () => {
      const opts = { opts: "my custom option" }
      const app = new App()
      const subApp = new App()
      const arr: Array<number> = [];
      app.setInit('init', async (instance, opts) => {
        arr.push(1);
      })
      app.addApp(subApp, opts)
      await app.start()

      expect(arr).toEqual([1]);
      expect(subApp.opts).toEqual(opts)
    })

    it("Just name and app", async () => {
      const app = new App()
      const subApp = new App()
      const arr: Array<number> = [];
      app.setInit('init', async (instance, opts) => {
        arr.push(1);
      })
      app.addApp("subApp", subApp)
      await app.start()

      expect(arr).toEqual([1]);
    })

    it("Name, App and Opts", async () => {
      const opts = { opts: "my custom option" }
      const app = new App()
      const subApp = new App()
      const arr: Array<number> = [];
      app.setInit('init', async (instance, opts) => {
        arr.push(1);
      })
      app.addApp("subApp", subApp, opts)
      await app.start()

      expect(arr).toEqual([1]);
      expect(subApp.opts).toEqual(opts)

    })
  })


});
