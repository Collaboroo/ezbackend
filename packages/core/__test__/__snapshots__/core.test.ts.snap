// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Behaviour An app should not be able to have 2 parents 1`] = `[Error: App's parent has already been set, parents can only be set once]`;

exports[`Default Behaviour Hooks Removing a non-existent hook should throw an error 1`] = `[Error: non-existent does not exist in _handler]`;

exports[`Default Behaviour Hooks Setting a hook twice should throw an error 1`] = `[Error: Set Flag already declared for _init]`;

exports[`Default Behaviour Two child apps should not have the same name 1`] = `[Error: Child app child already exists]`;

exports[`Default Behaviour addApp syntax must be strictly followed 1`] = `[Error: Invalid function signature for addApp()]`;
